File: Lab7-ListDemo/Lab7-ListDemo/Lab7-ListDemo/TableViewDemoVC/TableViewDemoVC.m
//
//  TableViewDemoVC.m
//  Lab7-ListDemo
//
//  Created by 張力元 on 2019/3/4.
//  Copyright © 2019 張力元. All rights reserved.
//

#import "TableViewDemoVC.h"
#import "MyTableViewCell.h"

@interface TableViewDemoVC ()<UITableViewDelegate,UITableViewDataSource>
@property (strong, nonatomic) NSArray *fruitNameArr;
@end

@implementation TableViewDemoVC

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    _fruitNameArr = [userDefaults arrayForKey:@"friutName"];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    //設定每個Section有幾筆資料要顯示
    if (section == 0)
        return _fruitNameArr.count;
    return 20;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    //設定每個Cell要顯示的資料與畫面
    //Cell的重複利用機制 可以節省記憶體 將移出畫面的Cell移至即將移入畫面的Cell做使用
    //reuse ID需與init時相同
    if (indexPath.section == 0) {
        MyTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"MyTableViewCell"];
        if (cell == nil) {
            //若沒有可用的Cell 則初始化一個Cell並設定reuse ID
            cell = [[[NSBundle mainBundle] loadNibNamed:@"MyTableViewCell" owner:nil options:nil] objectAtIndex:0];
        }
        NSString *friut = _fruitNameArr[indexPath.row];
        [cell setCellImage:friut title:friut];
        return cell;
    }
    else{
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"reuseCell"];
        NSLog(@"%@",cell);
        if (cell == nil) {
            //若沒有可用的Cell 則初始化一個Cell並設定reuse ID
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"reuseCell"];
        }
        //設定Cell的內容
        cell.textLabel.text = [NSString stringWithFormat:@"第%d個Cell",(int)indexPath.row];
        //輸出設定完成的Cell
        return cell;
    }
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    //設定要幾個Section  若不設定則為1
    return 2;
}

- (nullable NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section{
    //設定section上方的title
    if (section == 0) {
        return @"客製化Table View Cell";
    }
    return @"系統內建的Table View Cell";
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section{
    //設定section上方的高度
    return 50;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    //點擊某個Cell執行的動作
    NSLog(@"你點擊了 %d-%d 的Cell",(int)indexPath.section,(int)indexPath.row);
    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
    NSLog(@"Cell 的高為 %d",(int)cell.bounds.size.height);
    //取消選取的動畫
    [tableView deselectRowAtIndexPath:indexPath animated:true];
    
}
@end

File: Swift/lab7/lab7_exercise/lab7_exercise/tableViewDemoVC/tableViewDemoVC.swift
//
//  tableViewDemoVC.swift
//  lab7-1-2
//
//  Created by taizhou on 2019/7/29.
//  Copyright © 2019 taizhou. All rights reserved.
//

import UIKit

class tableViewDemoVC: UIViewController {

    @IBOutlet var tableView: UITableView!
    var fruitNameArr : [String]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        //初始化Table View
        tableViewInit()
    }
     
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        //初始化水果名稱的陣列，使用NSUserDefaults呼叫之前設定的陣列
        let userDefault = UserDefaults.standard
        fruitNameArr = userDefault.value(forKey: "fruitName") as? [String]
    }
    
    func tableViewInit(){
        //可參考7-1-1的「註冊Table View Cell」
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "reuseCell")
        let cellNib = UINib(nibName: "myTableViewCell", bundle: nil)
        tableView.register(cellNib, forCellReuseIdentifier: "myTableViewCell")
    }
}

extension tableViewDemoVC:UITableViewDataSource{
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if(section == 0 ){
            return fruitNameArr?.count ?? 0
        }
        else{
            return 20
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if(indexPath.section == 0 ){
            let cell = tableView.dequeueReusableCell(withIdentifier: "myTableViewCell", for: indexPath) as! myTableViewCell
            let fruit = fruitNameArr?[indexPath.row]
            cell.setCell(imgName: fruit ?? "", title: fruit ?? "")
            return cell
        }
        else{
            let cell = tableView.dequeueReusableCell(withIdentifier: "reuseCell", for: indexPath)
            cell.textLabel?.text = String(format:"第%d個cell",indexPath.row)
            return cell
        }
    }
    
    func numberOfSections(in tableView: UITableView) -> Int{
        return 2
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String?{
        if(section == 0 ){
            return "客製化 table view cell"
        }
        else{
            return "系統內建 table view cell"
        }
    }
}

extension tableViewDemoVC:UITableViewDelegate{
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat{
        //設定section上方的高度
        return 50
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath){
        //使用print()顯示出被點擊Cell的小節（section）與項目序號（row）
        print("你點擊了",indexPath.section,"-",indexPath.row)
        //讀取Table View Cell，並使用print()顯示出Cell的高度
        let cell = tableView.cellForRow(at: indexPath)!
        print("Cell的高度為"+String(Int(cell.bounds.size.height)))
        //取消選取的動畫
        tableView.deselectRow(at: indexPath, animated: true)
    }
}
